File: FCL/src/main/java/com/tungsten/fcl/ui/account/OfflineAccountSkinDialog.java
package com.tungsten.fcl.ui.account;

import static com.tungsten.fcl.game.TexturesLoader.getDefaultSkin;
import static com.tungsten.fclcore.util.Logging.LOG;

import android.app.Activity;
import android.content.Context;
import android.graphics.Point;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.tungsten.fcl.R;
import com.tungsten.fcl.activity.MainActivity;
import com.tungsten.fcl.util.FXUtils;
import com.tungsten.fcl.util.RequestCodes;
import com.tungsten.fclcore.auth.offline.OfflineAccount;
import com.tungsten.fclcore.auth.offline.Skin;
import com.tungsten.fclcore.auth.yggdrasil.TextureModel;
import com.tungsten.fclcore.fakefx.beans.InvalidationListener;
import com.tungsten.fclcore.fakefx.beans.property.ObjectProperty;
import com.tungsten.fclcore.fakefx.beans.property.SimpleObjectProperty;
import com.tungsten.fclcore.task.Schedulers;
import com.tungsten.fclcore.util.StringUtils;
import com.tungsten.fcllibrary.browser.FileBrowser;
import com.tungsten.fcllibrary.browser.options.LibMode;
import com.tungsten.fcllibrary.browser.options.SelectionMode;
import com.tungsten.fcllibrary.component.dialog.FCLDialog;
import com.tungsten.fcllibrary.component.view.FCLButton;
import com.tungsten.fcllibrary.component.view.FCLEditText;
import com.tungsten.fcllibrary.component.view.FCLImageButton;
import com.tungsten.fcllibrary.component.view.FCLLinearLayout;
import com.tungsten.fcllibrary.component.view.FCLRadioButton;
import com.tungsten.fcllibrary.component.view.FCLTextView;
import com.tungsten.fcllibrary.skin.SkinCanvas;
import com.tungsten.fcllibrary.skin.SkinRenderer;
import com.tungsten.fcllibrary.util.ConvertUtils;

import java.util.ArrayList;
import java.util.logging.Level;

public class OfflineAccountSkinDialog extends FCLDialog implements View.OnClickListener {

    private final AccountListItem accountListItem;
    private final OfflineAccount account;

    private ConstraintLayout root;
    private SkinCanvas skinCanvas;
    private SkinRenderer renderer;

    private FCLTextView title;
    private FCLButton positive;
    private FCLButton negative;

    private FCLRadioButton defaultSkin;
    private FCLRadioButton steve;
    private FCLRadioButton alex;
    private FCLRadioButton local;
    private FCLRadioButton csl;

    private FCLLinearLayout localLayout;
    private FCLLinearLayout cslLayout;
    private FCLImageButton skinPath;
    private FCLImageButton capePath;
    private FCLTextView skinPathText;
    private FCLTextView capePathText;
    private FCLEditText cslUrl;

    private final InvalidationListener skinBinding;
    private final ObjectProperty<Skin.Type> typeProperty = new SimpleObjectProperty<>(this, "type", Skin.Type.DEFAULT);

    public OfflineAccountSkinDialog(@NonNull Context context, AccountListItem accountListItem) {
        super(context);
        this.accountListItem = accountListItem;
        this.account = (OfflineAccount) accountListItem.getAccount();

        WindowManager wm = getWindow().getWindowManager();
        Point point = new Point();
        wm.getDefaultDisplay().getSize(point);
        WindowManager.LayoutParams params = getWindow().getAttributes();
        params.width = point.x * 2 / 3;
        if (point.y * 2 < point.x) {
            params.height = WindowManager.LayoutParams.MATCH_PARENT;
        } else {
            params.height = point.y * 2 / 3;
        }
        getWindow().setAttributes(params);

        setContentView(R.layout.dialog_offline_account_skin);
        setCancelable(false);

        renderer = new SkinRenderer(getContext());
        skinCanvas = findViewById(R.id.skin_view);
        skinCanvas.setRenderer(renderer, 5f);

        root = findViewById(R.id.root);
        title = findViewById(R.id.title);

        defaultSkin = findViewById(R.id.default_skin);
        steve = findViewById(R.id.steve);
        alex = findViewById(R.id.alex);
        local = findViewById(R.id.local);
        csl = findViewById(R.id.csl);

        defaultSkin.setOnClickListener(this);
        steve.setOnClickListener(this);
        alex.setOnClickListener(this);
        local.setOnClickListener(this);
        csl.setOnClickListener(this);

        localLayout = findViewById(R.id.local_layout);
        cslLayout = findViewById(R.id.csl_layout);

        skinPath = findViewById(R.id.skin_path);
        capePath = findViewById(R.id.cape_path);
        skinPath.setOnClickListener(this);
        capePath.setOnClickListener(this);
        skinPathText = findViewById(R.id.skin_path_text);
        capePathText = findViewById(R.id.cape_path_text);

        cslUrl = findViewById(R.id.csl_url);

        positive = findViewById(R.id.positive);
        negative = findViewById(R.id.negative);
        negative.post(() -> {
            positive.setOnClickListener(this);
            negative.setOnClickListener(this);
        });

        if (account.getSkin() == null) {
            refreshRadio(0);
            typeProperty.set(Skin.Type.DEFAULT);
        } else {
            if (account.getSkin().getType() == Skin.Type.STEVE) {
                refreshRadio(1);
                typeProperty.set(Skin.Type.STEVE);
            } else if (account.getSkin().getType() == Skin.Type.ALEX) {
                refreshRadio(2);
                typeProperty.set(Skin.Type.ALEX);
            } else if (account.getSkin().getType() == Skin.Type.LOCAL_FILE) {
                refreshRadio(3);
                typeProperty.set(Skin.Type.LOCAL_FILE);
            } else if (account.getSkin().getType() == Skin.Type.CUSTOM_SKIN_LOADER_API) {
                refreshRadio(4);
                typeProperty.set(Skin.Type.CUSTOM_SKIN_LOADER_API);
            } else {
                refreshRadio(0);
                typeProperty.set(Skin.Type.DEFAULT);
            }
            skinPathText.setString(account.getSkin().getLocalSkinPath());
            capePathText.setString(account.getSkin().getLocalCapePath());
            cslUrl.setText(account.getSkin().getCslApi());
        }

        refreshSkin();
        skinBinding = FXUtils.observeWeak(this::refreshSkin, typeProperty, cslUrl.stringProperty(), skinPathText.stringProperty(), capePathText.stringProperty());
    }

    private void refreshSkin() {
        getSkin().load(account.getUsername())
                .whenComplete(Schedulers.androidUIThread(), (result, exception) -> {
                    if (exception != null) {
                        LOG.log(Level.WARNING, "Failed to load skin", exception);
                        Toast.makeText(getContext(), getContext().getString(R.string.message_failed), Toast.LENGTH_SHORT).show();
                    } else {
                        if (result == null || result.getSkin() == null && result.getCape() == null) {
                            renderer.setTexture(getDefaultSkin(TextureModel.detectUUID(account.getUUID())).getImage(), null);
                            return;
                        }
                        renderer.setTexture(result.getSkin() != null ? result.getSkin().getImage() : getDefaultSkin(TextureModel.detectUUID(account.getUUID())).getImage(),
                                result.getCape() != null ? result.getCape().getImage() : null);
                    }
                }).start();
    }

    private void refreshRadio(int position) {
        defaultSkin.setChecked(position == 0);
        steve.setChecked(position == 1);
        alex.setChecked(position == 2);
        local.setChecked(position == 3);
        csl.setChecked(position == 4);

        localLayout.setVisibility(position == 3 ? View.VISIBLE : View.GONE);
        cslLayout.setVisibility(position == 4 ? View.VISIBLE : View.GONE);
    }

    private Skin getSkin() {
        return new Skin(typeProperty.get(), cslUrl.getStringValue() == null ? "" : cslUrl.getStringValue(), null, StringUtils.isBlank(skinPathText.getString()) ? null : skinPathText.getString(), StringUtils.isBlank(capePathText.getString()) ? null : capePathText.getString());
    }

    @Override
    public void onClick(View view) {
        if (view == defaultSkin) {
            refreshRadio(0);
            typeProperty.set(Skin.Type.DEFAULT);
        }
        if (view == steve) {
            refreshRadio(1);
            typeProperty.set(Skin.Type.STEVE);
        }
        if (view == alex) {
            refreshRadio(2);
            typeProperty.set(Skin.Type.ALEX);
        }
        if (view == local) {
            refreshRadio(3);
            typeProperty.set(Skin.Type.LOCAL_FILE);
        }
        if (view == csl) {
            refreshRadio(4);
            typeProperty.set(Skin.Type.CUSTOM_SKIN_LOADER_API);
        }

        if (view == skinPath) {
            FileBrowser.Builder builder = new FileBrowser.Builder(getContext());
            builder.setLibMode(LibMode.FILE_CHOOSER);
            builder.setSelectionMode(SelectionMode.SINGLE_SELECTION);
            ArrayList<String> suffix = new ArrayList<>();
            suffix.add(".png");
            builder.setSuffix(suffix);
            builder.create().browse(MainActivity.getInstance(), RequestCodes.SELECT_SKIN_CODE, ((requestCode, resultCode, data) -> {
                if (requestCode == RequestCodes.SELECT_SKIN_CODE && resultCode == Activity.RESULT_OK && data != null) {
                    String path = FileBrowser.getSelectedFiles(data).get(0);
                    skinPathText.setString(path);
                }
            }));
        }
        if (view == capePath) {
            FileBrowser.Builder builder = new FileBrowser.Builder(getContext());
            builder.setLibMode(LibMode.FILE_CHOOSER);
            builder.setSelectionMode(SelectionMode.SINGLE_SELECTION);
            ArrayList<String> suffix = new ArrayList<>();
            suffix.add(".png");
            builder.setSuffix(suffix);
            builder.create().browse(MainActivity.getInstance(), RequestCodes.SELECT_CAPE_CODE, ((requestCode, resultCode, data) -> {
                if (requestCode == RequestCodes.SELECT_CAPE_CODE && resultCode == Activity.RESULT_OK && data != null) {
                    String path = FileBrowser.getSelectedFiles(data).get(0);
                    capePathText.setString(path);
                }
            }));
        }

        if (view == positive) {
            account.setSkin(getSkin());
            accountListItem.refreshSkinBinding();
            dismiss();
        }
        if (view == negative) {
            dismiss();
        }
    }
}

File: FCL/src/main/java/com/tungsten/fcl/ui/account/OfflineAccountSkinDialog.kt
package com.tungsten.fcl.ui.account

import android.content.Context
import android.view.View
import android.view.WindowManager
import android.widget.Toast
import com.tungsten.fcl.R
import com.tungsten.fcl.activity.MainActivity
import com.tungsten.fcl.databinding.DialogOfflineAccountSkinBinding
import com.tungsten.fcl.game.TexturesLoader
import com.tungsten.fclcore.auth.offline.OfflineAccount
import com.tungsten.fclcore.auth.offline.Skin
import com.tungsten.fclcore.auth.offline.Skin.LoadedSkin
import com.tungsten.fclcore.auth.yggdrasil.TextureModel
import com.tungsten.fclcore.task.Schedulers
import com.tungsten.fclcore.util.Logging
import com.tungsten.fclcore.util.StringUtils
import com.tungsten.fclauncher.utils.FCLPath
import com.tungsten.fcllibrary.component.dialog.FCLDialog
import com.tungsten.fcllibrary.component.view.FCLTextView
import com.tungsten.fcllibrary.skin.SkinRenderer
import java.io.File
import java.util.logging.Level
import com.mio.util.getScreenHeight
import com.mio.util.getScreenWidth

class OfflineAccountSkinDialog(context: Context, private val accountListItem: AccountListItem) :
    FCLDialog(context), View.OnClickListener {
    private val account: OfflineAccount = accountListItem.account as OfflineAccount
    private var binding: DialogOfflineAccountSkinBinding =
        DialogOfflineAccountSkinBinding.inflate(layoutInflater)
    private val renderer: SkinRenderer

    /** Texture model chosen via the classic/slim RadioGroup */
    private var model: TextureModel = TextureModel.STEVE

    init {
        setContentView(binding.root)
        setCancelable(false)

        renderer = SkinRenderer(getContext())
        binding.skinView.setRenderer(renderer, 5f)

        binding.modelClassic.setOnClickListener(this)
        binding.modelSlim.setOnClickListener(this)
        binding.skinPath.setOnClickListener(this)
        binding.skinPathReset.setOnClickListener(this)
        binding.capePath.setOnClickListener(this)
        binding.capePathReset.setOnClickListener(this)
        binding.positive.setOnClickListener(this)
        binding.negative.setOnClickListener(this)

        // Restore state from existing skin
        val skin = account.skin
        if (skin == null) {
            model = TextureModel.detectUUID(account.uuid)
        } else {
            // Restore model: legacy ALEX -> slim, otherwise from stored texture model
            model = if (skin.type() == Skin.Type.ALEX) TextureModel.ALEX else skin.textureModel()
            binding.skinPathText.string = skin.localSkinPath()
            binding.capePathText.string = skin.localCapePath()
        }
        refreshModelRadio()
    }

    override fun show() {
        val width = getScreenWidth()
        var height = getScreenHeight()
        height = if (height * 2 < width) {
            WindowManager.LayoutParams.MATCH_PARENT
        } else {
            height * 2 / 3
        }
        window?.setLayout(width * 2 / 3, height)
        super.show()
        binding.skinView.onResume()
        refreshSkin()
    }

    override fun dismiss() {
        binding.skinView.onPause()
        super.dismiss()
    }

    private fun refreshSkin() {
        this.skin.load()
            .whenComplete(
                Schedulers.androidUIThread()
            ) { result: LoadedSkin?, exception: Exception? ->
                if (exception != null) {
                    Logging.LOG.log(Level.WARNING, "Failed to load skin", exception)
                    Toast.makeText(
                        context,
                        context.getString(R.string.message_failed),
                        Toast.LENGTH_SHORT
                    ).show()
                } else {
                    // Model is controlled by the user's classic/slim selection
                    val slim = model == TextureModel.ALEX
                    if (result == null || result.skin() == null && result.cape() == null) {
                        // No custom skin selected: fall back to the default skin of the chosen model
                        renderer.updateTexture(
                            TexturesLoader.getDefaultSkin(model).image(),
                            null,
                            slim
                        )
                        return@whenComplete
                    }
                    renderer.updateTexture(
                        if (result.skin() != null) result.skin()
                            .image else TexturesLoader.getDefaultSkin(
                            model
                        ).image(),
                        if (result.cape() != null) result.cape().image else null,
                        slim
                    )
                }
            }.start()
    }

    private fun refreshModelRadio() {
        binding.modelClassic.setChecked(model == TextureModel.STEVE)
        binding.modelSlim.setChecked(model == TextureModel.ALEX)
    }

    private val skin: Skin
        get() {
            val skinPath = binding.skinPathText.string
            val capePath = binding.capePathText.string
            val hasSkin = StringUtils.isNotBlank(skinPath)
            val hasCape = StringUtils.isNotBlank(capePath)
            return if (hasSkin || hasCape) {
                Skin(
                    Skin.Type.LOCAL_FILE,
                    model,
                    if (hasSkin) skinPath else null,
                    if (hasCape) capePath else null
                )
            } else {
                Skin(Skin.Type.DEFAULT, model, null, null)
            }
        }

    override fun onClick(view: View?) {
        when (view) {
            // Model selection
            binding.modelClassic -> {
                model = TextureModel.STEVE
                refreshModelRadio()
                refreshSkin()
            }

            binding.modelSlim -> {
                model = TextureModel.ALEX
                refreshModelRadio()
                refreshSkin()
            }
            // File selection：先放缓存，点确认时才落位 SKIN_DIR，取消不影响已有皮肤文件
            binding.skinPath -> MainActivity.getInstance().fileLauncher.launchSingleSelection(
                null,
                listOf(".png")
            ) {
                binding.skinPathText.string =
                    it?.get(0)?.toFile(context, File(FCLPath.CACHE_DIR))?.absolutePath
                        ?: return@launchSingleSelection
                refreshSkin()
            }

            binding.skinPathReset -> {
                binding.skinPathText.string = null
                refreshSkin()
            }

            binding.capePath -> MainActivity.getInstance().fileLauncher.launchSingleSelection(
                null,
                listOf(".png")
            ) {
                binding.capePathText.string =
                    it?.get(0)?.toFile(context, File(FCLPath.CACHE_DIR))?.absolutePath
                        ?: return@launchSingleSelection
                refreshSkin()
            }

            binding.capePathReset -> {
                binding.capePathText.string = null
                refreshSkin()
            }
            // Buttons
            binding.positive -> {
                commitSkinFile(binding.skinPathText, "${account.uuid}.png")
                commitSkinFile(binding.capePathText, "${account.uuid}_cape.png")
                account.setSkin(this.skin)
                accountListItem.refreshSkinBinding()
                dismiss()
            }

            binding.negative -> dismiss()
        }
    }

    /** 把临时选中的皮肤/披风文件落位到 SKIN_DIR 固定名，并更新为最终路径 */
    private fun commitSkinFile(text: FCLTextView, fileName: String) {
        val srcPath = text.string ?: return
        val src = File(srcPath)
        val dest = File(FCLPath.SKIN_DIR, fileName)
        if (src.canonicalPath != dest.canonicalPath) {
            runCatching { src.copyTo(dest, overwrite = true) }
                .onSuccess { text.string = dest.absolutePath }
        }
    }
}
